src/components/Map.jsx

```diff
@@ -33,17 +68,95 @@ function Map() {
   // if there are layers, calculate the bounds
   // each layer is a feature collection, so we need to calculate the bounds of all features
   useEffect(() => {
+    if (!mapRef.current) return;
+    const map = mapRef.current.getMap();
+
     if (mapLayers.length > 0) {
+
+      // Dynamically add map layers
+      mapLayers.forEach((layer, idx) => {
+        const layerId = layer.id || idx;
+        const isVisible = layerVisibility[layerId] ?? true;
+
+        if (layer.type === "geojson") {
+          if (map.getSource(layerId)) {
+            map.getSource(layerId).setData(layer.data); // refresh the data
+          }
+          else {
+            map.addSource(layerId, {
+              type: "geojson",
+              data: layer.data,
+            });
+
+            // Add the fill layer
+            map.addLayer(
+              {
+                id: `fill-layer-${layerId}`,
+                type: "fill",
+                source: layerId,
+              },
+            );
+
+            // Add the line layer
+            map.addLayer(
+              {
+                id: `line-layer-${layerId}`,
+                type: "line",
+                source: layerId,
+                paint: {
+                  "line-width": 2
+                },
+              },
+            );
+          }
+
+            // set the visibility of the layer
+            map.setLayoutProperty(`fill-layer-${layerId}`, "visibility", isVisible ? "visible" : "none");
+            map.setLayoutProperty(`line-layer-${layerId}`, "visibility", isVisible ? "visible" : "none");
+
+            // set paint properties
+            map.setPaintProperty(`fill-layer-${layerId}`, "fill-color", ["case", 
+                    ["all",
+                      ["has", "gadm_id"],
+                      ["==", ["get", "gadm_id"], highlightedLocation ?? null],
+                    ], pink500, blue500
+                  ]);
+            map.setPaintProperty(`fill-layer-${layerId}`, "fill-opacity", [
+                    "case",
+                    ["all",
+                      ["has", "gadm_id"],
+                      ["==", ["get", "gadm_id"], highlightedLocation ?? null],
+                    ], 0.5, 0.25
+                  ]);
+            map.setPaintProperty(`line-layer-${layerId}`, "line-color", ["case",
+                    ["all",
+                      ["has", "gadm_id"],
+                      ["==", ["get", "gadm_id"], highlightedLocation ?? null],
+                    ], pink500, blue500
+                  ]
+            );
+        }
+        if (layer.type === "TMS") {
+          // Add or update TMS layer
+          addTMSLayer(map, layerId, layer.url, isVisible);
+
+        }
+      });
```

Comment:
Same here, managing the layers in the effect hook might not be necessary with ReactMapGL.